lib/saluki-env/src/workload/collectors/cgroups_v2.rs

```diff
@@ -0,0 +1,153 @@
+        builder.firm().with_fixed_amount(std::mem::size_of::<Self>());
+    }
+}
+
+fn traverse_cgroups(
+    root: &CgroupReader, interner: &FixedSizeInterner<1>, operations: &mut Vec<MetadataOperation>,
+) -> Result<(), GenericError> {
+    for child_cgroup in root.child_cgroup_iter()? {
+        let cgroup_name = child_cgroup.name().as_os_str().to_string_lossy();
+
+        // Check if this control group is associated with a container.
+        if let Some(container_id) = extract_container_id(&cgroup_name, interner) {
+            debug!(container_id = %container_id, %cgroup_name, "Found container control group.");
+
+            // Get the inode for this control group.
+            let cgroup_inode = child_cgroup.read_inode_number()?;
+
+            // Create an ancestry link between the container inode and the container ID.
+            let entity_id = EntityId::ContainerInode(cgroup_inode);
+            let ancestor_entity_id = EntityId::Container(container_id);
+
+            let operation = MetadataOperation::link_ancestor(entity_id, ancestor_entity_id);
+            operations.push(operation);
+        }
+
+        // After that, traverse the children of this control group.
+        traverse_cgroups(&child_cgroup, interner, operations)?;
+    }
+
+    Ok(())
+}
+
+fn extract_container_id(cgroup_name: &str, interner: &FixedSizeInterner<1>) -> Option<MetaString> {
+    // This regular expression is meant to capture:
+    // - 64 character hexadecimal strings (standard format for container IDs almost everywhere)
+    // - 32 character hexadecimal strings followed by a dash and a number (used by AWS ECS)
+    // - 8 character hexadecimal strings followed by up to four groups of 4 character hexadecimal strings separated by
+    //   dashes (essentially a UUID, used by Pivotal Cloud Foundry's Garden technology)
+    static CONTAINER_REGEX: LazyLock<Regex> =
+        LazyLock::new(|| Regex::new("([0-9a-f]{64})|([0-9a-f]{32}-\\d+)|([0-9a-f]{8}(-[0-9a-f]{4}){4}$)").unwrap());
+
+    match CONTAINER_REGEX.find(cgroup_name) {
+        Some(name) => {
+            // NOTE: We've lifted this logic from the Datadog Agent [1] to handle filtering out certain control groups
+            // based on how systemd names them.
+            //
+            // [1]: https://github.com/DataDog/datadog-agent/blob/fe75b815c2f135f0d2ea85d7a57a8fc8cbf56bd9/pkg/util/cgroups/reader.go#L63-L77
+            if name.as_str().ends_with(".mount") || name.as_str().starts_with("crio-conmon-") {
+                None
+            } else {
+                match interner.try_intern(name.as_str()) {
+                    Some(interned) => Some(MetaString::from(interned)),
+                    None => {
+                        error!(container_id = %name.as_str(), "Failed to intern container ID.");
+                        None
+                    }
+                }
+            }
+        }
+        None => None,
```

Comment:
You could probably use `Option::filter` in place of the outer outer match and first conditional to make this a little more concise.

Reply:
Good call. Cleaned this up a little in 8b9038c.